Add item requests

VitalyKapustin-git opened the pull request:
- added functionallity for item request;
- added tests;

VitalyKapustin-git committed:
feat:
- implemented requests;

VitalyKapustin-git committed:
feat:
- implemented unit tests;

VitalyKapustin-git committed:
feat:
- added tests;

VitalyKapustin-git committed:
feat:
- added tests;

NikolaiDem commented on src/main/java/ru/practicum/shareit/booking/dao/BookingRepository.java:
```diff
@@ -1,59 +1,67 @@
 import java.time.LocalDateTime;
 import java.util.List;
 
+@Repository
```
Аннотация не нужна, этот интерфейс бином не станет)

NikolaiDem commented on src/main/java/ru/practicum/shareit/booking/mapper/BookingMapper.java:
```diff
@@ -1,31 +1,18 @@
         bookingDto.setId(booking.getId());
         bookingDto.setStart(booking.getStart());
         bookingDto.setEnd(booking.getEnd());
-        bookingDto.setItem(ItemMapper.toItemDto(itemRepository.getItemById(booking.getItemId())));
```
Теперь маппер стал не совсем полноценным. Часть маппинга выполняется здесь, а часть в других местах. Лучше сделать поля Item и User в Booking-е и смапить их на ItemDto и UserDto в BookingDto здесь.

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -31,4 +31,7 @@ public class Item {
     private long ownerId;
 
+    @Column(name = "request_id")
+    private Long requestId;
```
Так как у запроса есть описание, заказчик и другие параметры, то удобнее работать с объектом ItemRequest. Поэтому вместо requestId лучше создать ItemRequest.

VitalyKapustin-git replied:
@NikolaiDem Этот момент остался единственный без правок. 
Слишком много логики переписывать, концов с концами не сыщешь.
Если можно, этот момент я бы уже оставил как есть, с учетом того, что остальные требования удовлетворены.

NikolaiDem replied:
Так как используется только в маппере. оставим так.

NikolaiDem commented on src/main/java/ru/practicum/shareit/request/model/ItemRequest.java:
```diff
@@ -0,0 +1,32 @@
+import java.time.LocalDateTime;
+
+/**
+ * TODO Sprint add-item-requests.
```
TODO пишем, когда есть что делать, а здесь все готово)

NikolaiDem commented on src/main/java/ru/practicum/shareit/request/model/ItemRequest.java:
```diff
@@ -0,0 +1,32 @@
+@Entity
+@Table(name = "requests")
+@NoArgsConstructor
+public class ItemRequest {
```
Все поля private

NikolaiDem commented on src/main/java/ru/practicum/shareit/request/model/ItemRequest.java:
```diff
@@ -0,0 +1,32 @@
+    String description;
+
+    @Column(name = "requestor_id")
+    Long requestorId;
```
Лучше использовать User вместо requestorId.

NikolaiDem commented on src/main/java/ru/practicum/shareit/request/controller/ItemRequestController.java:
```diff
@@ -0,0 +1,48 @@
+    ItemRequestService itemRequestService;
+
+    @PostMapping
+    public ItemRequestDto create(@Valid @RequestBody ItemRequest itemRequest,
```
При общении с внешним миром лучше использовать структуры DTO (ItemRequestDto)

NikolaiDem commented on src/main/java/ru/practicum/shareit/request/dto/ItemRequestDto.java:
```diff
@@ -1,7 +1,26 @@
+@Data
 public class ItemRequestDto {
+
+    Long id;
```
Все поля private

NikolaiDem commented on src/main/java/ru/practicum/shareit/request/dto/ItemRequestDto.java:
```diff
@@ -1,7 +1,26 @@
+
+    String description;
+
+    Long requestorId;
```
Это поле не используется ни для запроса ни для ответа)

NikolaiDem commented on src/main/java/ru/practicum/shareit/request/service/ItemRequestService.java:
```diff
@@ -0,0 +1,18 @@
+
+public interface ItemRequestService {
+
+    public ItemRequestDto create(ItemRequest itemRequest, Long requestorId);
```
В интерфейсе все методы по умолчанию public. Явно указывать не нужно.

NikolaiDem commented on src/main/java/ru/practicum/shareit/request/model/ItemRequest.java:
```diff
@@ -0,0 +1,32 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    long id;
+
+    @NotBlank
```
Валидацию лучше делать на классах DTO, потому что контроллеры обмениваются через них с внешним миром.

NikolaiDem commented on src/main/java/ru/practicum/shareit/request/service/ItemRequestServiceImpl.java:
```diff
@@ -0,0 +1,115 @@
+
+    @Transactional
+    @Override
+    public ItemRequestDto create(ItemRequest itemRequest, Long requestorId) {
```
Здесь должно быть ItemRequestDto, потому и проверка if (itemRequest.getDescription().isBlank()){} будет лишней, потому что невалидный Dto не пройдет контроллер.

VitalyKapustin-git replied:
Здесь проверка не для того чтобы проверить, а для того, чтобы выкинуть свой exception для прохождения тестов в постмане с человеческим текстом).
Убрал.

NikolaiDem commented on src/main/java/ru/practicum/shareit/request/service/ItemRequestServiceImpl.java:
```diff
@@ -0,0 +1,115 @@
+        return itemRequestRepository.getItemRequestsByRequestorId(userId).stream()
+                .map(v -> {
+                    ItemRequestDto itemRequestDto = ItemRequestMapper.toItemRequestDto(v);
+                    itemRequestDto.setItems(itemRepository.getItemsByRequestId(v.getId()).stream()
```
Здесь выполнятся лишние запросы к БД. Лучше одним запросом вытащить все Item-s для всех запросов, а уже здесь распределить их по ItemRequest-ам.
Аналогично для getOtherUsersRequests(...)

NikolaiDem commented on src/main/resources/schema.sql:
```diff
@@ -1,48 +1,201 @@
-);
+CREATE TABLE IF NOT EXISTS users
+(
+    id
```
Привести к читаемому виду (каждый столбец на своей строке)

VitalyKapustin-git replied:
@NikolaiDem Привет. Вот тут тебя не совсем понял. У меня же вроде так и есть? Или речь про добавление табуляций?

[скрин как у меня](https://imgur.com/a/bLUYzeM)

NikolaiDem replied:
Теперь все окей) Строки были разбиты по словам.

NikolaiDem commented on src/test/java/ru/practicum/shareit/booking/BookingServiceJPATest.java:
```diff
@@ -0,0 +1,188 @@
+@RequiredArgsConstructor(onConstructor_ = @Autowired)
+public class BookingServiceJPATest {
+
+    private final EntityManager em;
```
Так как тесты предназначаются для проверки работы репозитория, то лучше использовать его вместо менеджера сущностей. Проверка будет заключаться в том, чтобы сначала сохранить некоторые объекты, а потом вызывать кастомные запросы из репозитория и сравнить их результат с ожидаемым)

VitalyKapustin-git committed:
refactor:
- changed from type to typeDto in ItemRequestController;
- implemented @ManyToOne scheme in ItemRequest;
- implemented @ManyToOne scheme in ItemRequest;
- replaces EntityManager to repository implementation in JPA tests;
- optimized requests plan in ItemRequestServiceImpl;
- cleaned code;
-

NikolaiDem commented on src/main/java/ru/practicum/shareit/request/model/ItemRequest.java:
```diff
@@ -2,15 +2,11 @@
- * TODO Sprint add-item-requests.
- */
-
 @Data
```
Удалить @Data для всех сущностей. 
Методы equals и hashcode переопределять нет необходимости.

NikolaiDem commented on src/main/java/ru/practicum/shareit/request/service/ItemRequestServiceImpl.java:
```diff
@@ -101,15 +98,39 @@ public ItemRequestDto getRequest(Long requestId, Long userId) {
+    private void setItemsToItemRequest(List<ItemRequestDto> itemRequests) {
+        List<Long> itemRequestsId = itemRequests.stream().map(ItemRequestDto::getId).collect(Collectors.toList());
+
+        List<ItemDto> items = itemRepository.getItemsByRequestIdIn(itemRequestsId).stream()
```
Лучше в мапу преобразовать, чтобы избежать .filter(i -> i.getRequestId() == v.getId())

NikolaiDem commented on src/test/java/ru/practicum/shareit/item/ItemServiceUnitTest.java:
```diff
@@ -0,0 +1,323 @@
+package ru.practicum.shareit.item;
```
Добавить тест для ItemRepository на проверку кастомных запросов.

NikolaiDem commented on src/test/java/ru/practicum/shareit/booking/BookingControllerTest.java:
```diff
@@ -0,0 +1,153 @@
+package ru.practicum.shareit.booking;
```
Добавить тесты для мапперов.

VitalyKapustin-git replied:
Покрыл в юнит-тестах

VitalyKapustin-git committed:
feat:
- add tests for booking mappers to unit tests;
- add tests for custom queries from itemRepository in ItemServiceJPATest class;

refactor:
- refactored setItemsToItemRequest method in ItemRequestServiceImpl (done via Map implementation);
- replaces all @Data notations to @Setter and @Getter;
